pallets/proxy-keystore/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+                                return Err(Error::<T>::KeyWithHashExists.into())
+                            }
+                        },
+                        // No last key by purpose, we can continue.
+                        None => {},
+                    }
+
+                    // Replace any value that we might have in there since checks were OK.
+                    let _ = key_id_opt.insert(key_id.clone());
+
+                    Ok(())
+                },
+            )?;
+
+            // Insert the new key.
+            <Keys<T>>::insert(
+                account_id.clone(),
+                key_id,
+                Key{
+                    purpose: add_key.purpose.clone(),
+                    key_type: add_key.key_type.clone(),
+                    revoked_at: None,
+                    deposit: key_deposit,
+                },
+            );
+
+            Ok(())
+        }
+
+        /// Revoke a key with `key_hash` at the current `block_number` in the `Keys` storage
+        /// if the key is found and *not* revoked.
+        ///
+        /// Any entry that matches the `key_hash` and the key purpose in the `LastKeyByPurpose`
+        /// storage is removed.
+        ///
+        /// The key deposit is returned upon success.
+        fn revoke_key(
+            account_id: T::AccountId,
+            key_hash: T::Hash,
+            block_number: T::BlockNumber,
+        ) -> DispatchResult {
+            let mut key_id_opt: Option<KeyId<T::Hash>> = None;
+
+            let iter = <Keys<T>>::iter_prefix(account_id.clone());
+
+            // Search for a key that has the key_hash that we are looking for.
+            for (key_id, _) in iter {
+                if key_id.0 == key_hash {
+                    key_id_opt = Some(key_id);
+                    break
+                }
+            }
+
+            return match key_id_opt {
+                Some(key_id) => {
+                    // Retrieve the key from storage.
+                    // TODO(cdamian): Given the above, are there any chances of a race here?
+                    <Keys<T>>::try_mutate(
+                        account_id.clone(),
+                        key_id,
+                        |key_opt| -> DispatchResult  {
+                            return match key_opt {
+                                Some(key) => {
+                                    // Check if key was already revoked.
+                                    if let Some(_) = key.revoked_at {
+                                        return Err(Error::<T>::KeyRevoked.into())
+                                    }
+
+                                    // Revoke it at the current block number.
+                                    key.revoked_at = Some(block_number);
+
+                                    // Check if we have a key by purpose that matches our key_hash.
+                                    <LastKeyByPurpose<T>>::try_mutate(
```

Comment:
I believe we dont need to do anything here with the `LastKeyByPurpose` storage.

Reply:
Removed this line so we have an entry here even if the key is revoked.